src/compiler/transformers/esDecorators.ts

```diff
@@ -0,0 +1,2087 @@
+                            extendsExpression;
+                    classDefinitionStatements.push(createLet(classInfo.classSuper, safeExtendsExpression));
+                    const updatedExtendsElement = factory.updateExpressionWithTypeArguments(extendsElement, classInfo.classSuper, /*typeArguments*/ undefined);
+                    const updatedExtendsClause = factory.updateHeritageClause(extendsClause, [updatedExtendsElement]);
+                    heritageClauses = factory.createNodeArray([updatedExtendsClause]);
+                }
+            }
+            else {
+                // 2. ClassHeritage clause is evaluated outside of the private name scope of the class.
+                heritageClauses = visitNodes(node.heritageClauses, visitor, isHeritageClause);
+            }
+
+            const renamedClassThis = classInfo.classThis ?? factory.createThis();
+
+            // 3. The name of the class is assigned.
+            //
+            // If the class did not have a name, set the assigned name as if from NamedEvaluation.
+            // We don't need to use the assigned name if it consists of the empty string and the transformed class
+            // expression won't get its name from any other source (such as the variable we create to handle
+            // class decorators)
+            const needsSetNameHelper = !getOriginalNode(node, isClassLike)?.name && (classDecorators || !isStringLiteral(className) || !isEmptyStringLiteral(className));
+            if (needsSetNameHelper) {
+                const setNameExpr = emitHelpers().createSetFunctionNameHelper(factory.createThis(), className);
+                leadingBlockStatements = append(leadingBlockStatements, factory.createExpressionStatement(setNameExpr));
+            }
+
+            // 4. For each member:
+            //    a. Member Decorators are evaluated
+            //    b. Computed Property Name is evaluated, if present
+
+            // We visit members in two passes:
+            // - The first pass visits methods, accessors, and fields to collect decorators and computed property names.
+            // - The second pass visits the constructor to add instance initializers.
+            //
+            // NOTE: If there are no constructors, but there are instance initializers, a synthetic constructor is added.
+
+            enterClass(classInfo);
+            let members = visitNodes(node.members, classElementVisitor, isClassElement);
+            if (pendingExpressions) {
+                for (const expression of pendingExpressions) {
+                    const statement = factory.createExpressionStatement(expression);
+                    leadingBlockStatements = append(leadingBlockStatements, statement);
+                }
+                pendingExpressions = undefined;
+            }
+            exitClass();
+
+            if (classInfo.instanceExtraInitializersName && !getFirstConstructorWithBody(node)) {

```

Comment:
prepareConstructor checks `classInfo.instanceExtraInitializersName` and returns undefined if it's undefined so you don't technically have to here.

Reply:
It is a cheaper test than `getFirstConstructorWithBody(node)`. There's no sense looping over the class elements to find a constructor if we don't need to do the work in the first place.